### PR TITLE
GH-2755: Add DirectoryHashCalculator

### LINKED ISSUE #2755: Add DirectoryHashCalculator
Add a DirectoryHashCalculator in [Cake.Commen.Security](https://cakebuild.net/api/Cake.Common.Security/) for creating a hash of an given directory?

The usage would be about the same as the [FileHashCalculator](https://cakebuild.net/api/Cake.Common.Security/FileHashCalculator/), just that a DirectoryPath is specified instead of a [FilePath](https://cakebuild.net/api/Cake.Common.Security/FileHash/). The implementation should return a DirectoryHash instead of a FileHash. 

The private GetHashAlgorithm function in [FileHashCalculator](https://cakebuild.net/api/Cake.Common.Security/FileHashCalculator/) should be outsourced for common use.

### PR DESCRIPTION
**Pull request by eoehen**
- Add DirectoryHash represents a calculated directory hash.
- Add DirectoryHashCalculator for calculating a hash for a directory.
- Add HashAlgorithmBuilder for a common usage of creating a instance of the hash algorithm.

- Overload the ctor of FileHashCalculator with IHashAlgorithmBuilder as additional parameter.

Fixes #2755

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/cake-build/cake?pullRequest=2829) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/cake-build/cake?pullRequest=2829) before we can accept your contribution.<br/><br/>:x: eoehen [sign now](https://cla.dotnetfoundation.org/cake-build/cake?pullRequest=2829)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/cake-build/cake?pullRequest=2829) it.</sub>

File: src/Cake.Common/Security/HashAlgorithmBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    /// </summary>
+    public sealed class HashAlgorithmBuilder : IHashAlgorithmBuilder
+    {
+        /// <summary>
+        /// Return a instance of <see cref="System.Security.Cryptography.HashAlgorithm"/> by
+        /// <see cref="HashAlgorithm"/> enum value.
+        /// </summary>
+        /// <param name="hashAlgorithm">Algorithm enum value.</param>
+        /// <returns>A <see cref="System.Security.Cryptography.HashAlgorithm"/> instance.</returns>
+        public System.Security.Cryptography.HashAlgorithm GetHashAlgorithm(HashAlgorithm hashAlgorithm)
```
Comment (devlead) [inline src/Cake.Common/Security/HashAlgorithmBuilder.cs:18]:
```suggestion
        public System.Security.Cryptography.HashAlgorithm CreateHashAlgorithm(HashAlgorithm hashAlgorithm)
```
If this goes public I'll suggest we call it `Create`.

**Review by devlead — CHANGES_REQUESTED** [review]

**Comment by eoehen** [comment]
@devlead I have processed your feedback.

**Review by devlead — APPROVED** [review]
LGTM 👍

**Comment by devlead** [comment]
@eoehen your changes have been merged, thanks for your contribution 👍

**Pull request merged**